### PR TITLE
feat(cloudfront-origins): allow setting a user defined origin id

### LINKED ISSUE #2756: CloudFront Origin IDs
Hey there,
Currently we're migrating a serverless project to use the CDK. What we're experiencing now is that despite setting an ID for an Origin behavior it is always overwritten by `origin{indexOfBehavior+1}`. I also found that there is no typesafe way in TypeScript to set this property on `Behavior`, because it doesn't exist on that type. The parent construct `CloudFrontWebDistribution` is internally using an unexported type for behaviors called `BehaviorWithOrigin`. 
The constructor of `CloudFrontWebDistribution` is overwriting any specified (by sidestepping the typechecker) origin id, see [here](https://github.com/awslabs/aws-cdk/blob/e3df21a30160ec9d804c93951c9eb5975364d4ba/packages/@aws-cdk/aws-cloudfront/lib/web_distribution.ts#L550).

My question now is: is that the expected behavior of the CDK? Like, is this intentional to prevent people to have colliding origin IDs? Or is this a bug/feature?

### PR DESCRIPTION
**Pull request by corymhall**
This PR adds the ability to set a custom user defined origin identifier. The default identifier that is generated is not very descriptive and some users would like the ability to define a custom identifier.

Since ids have to be unique per distribution, I also added validation that will throw an error if it is not.

fixes #2756

----

### All Submissions:

* [ ] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [ ] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [ ] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by TheRealAmazonKendra** [comment]
@Mergifyio update

**Comment by TheRealAmazonKendra** [comment]
@corymhall why integ exempt?

**Comment by corymhall** [comment]
> @corymhall why integ exempt?

Our integ tests already set the `id` to some value that we generate. This change allows the user to define a different `id`. I don't think there is any _behavior_ to test in this case. We would be testing that when you set the `id` CloudFormation does the right thing, which is not something integ tests should test.

I also don't think this needs a README update, but let me know if you disagree.

File: packages/@aws-cdk/aws-cloudfront/lib/distribution.ts [hunk 1]
```diff
@@ -370,7 +374,7 @@ export class Distribution extends Resource implements IDistribution {
         this.addOriginGroup(originGroupId, originBindConfig.failoverConfig.statusCodes, originId, failoverOriginId);
         return originGroupId;
       }
-      return originId;
+      return originBindConfig.originProperty?.id ?? originId;
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-cloudfront/lib/distribution.ts:377]: Is this necessary? `originBindConfig.originProperty.id === originId` under our implementation of `OriginBase.bind()`.

I suppose someone could write their own `CustomOrigin extends IOrigin` and implement `bind` where the given `originId` does not become the `originBindConfig.originProperty.id`, but is that what we are defending against here?

If so... fine. But these will 99% of the time be the same value.

**Review by kaizencc — APPROVED** [review]
Approved with dnm for one minor nit comment. Feel free to resolve however you want.

File: packages/@aws-cdk/aws-cloudfront/lib/distribution.ts (hunk 1 above)
Comment (corymhall) [reply packages/@aws-cdk/aws-cloudfront/lib/distribution.ts:377]: They will be different if the user provides their own ID.

For example.

```ts
const origin = new S3Origin(bucket, { originId: 'MyCustomOrigin' });
```

When we then call bind on line `358` we pass in the generated originId (from line `357`. The change I made in `bind` is to take the originId passed in `S3Origin` first. So originBindConfig.originProperty.id will be `MyCustomOrigin`.

**Pull request merged**